drv/lpc55-spi-server/src/main.rs

```diff
@@ -63,74 +369,339 @@ fn main() -> ! {
+                if !spi.has_byte() {
+                    break;
+                }
+                let (b, csn, sot) = spi.read_u8_csn_sot(); // always read
+                ringbuf_entry!(Trace::Rx(rctx.state, b, rctx.count, csn, sot));
+                rctx.rx_byte(b, sot, &mut tctx);
+                if rctx.state == RxState::Header
+                    && tctx.state != TxState::Idle
+                    && tctx.rot_irq
+                {
+                    // Full duplex is not supported.
+                    // If the SP is sending a message to RoT
+                    // then clear out the TX state to make way for the
+                    // response to this message.
+                    gpio.set_val(rot_irq, Value::One).unwrap_lite();
+                    ringbuf_entry!(Trace::DeassertRotIrq);
+                    tctx.rot_irq = false;
+                    tctx.state = TxState::Idle;
+                    // drain Tx FIFO?
+                    // We will see as much as got put into the Tx FIFO go
+                    // over the SPI bus.
+                }
             }
 
+            // CSn de-assert is detected to catch the end of frame condition.
+            // If Rx end of frame occurs and no data is left in fifo to satisfy
+            // read length, then that is an error (an underrun that should have
+            // been already caught or a failure to send all the data.)
+            // End of frame disagreement with count from header is cause
+            // for sending an error response.
+            if !inframe && !spi.has_byte() {
+                // The SP is not talking to us, it cannot be consuming data
+                // If we were recieving and didn't get a complete message,
+                // then we're not getting any more bytes and that could be
+                // an error condition.
+                // If we were transmitting, then transmit cannot continue.
+
+                // partial header or payload received
+                if rctx.state == RxState::Header
+                    || rctx.state == RxState::Payload
+                {
+                    // XXX include a more useful error response?
+                    ringbuf_entry!(Trace::RxFragment);
+                    if tctx.enqueue(MsgType::ErrorRsp, EnqueueBuf::Empty) {
+                        ringbuf_entry!(Trace::CannotSendRxFragError);
+                        rctx.state = RxState::Dispatch;
+                    }
+                }
+                // Process transitory states
+                if rctx.state == RxState::Dispatch {
+                    // Receive is complete.
+                    // Processing may take a long time and detract from keeping up
+                    // with Tx and Rx FIFOs.
+                    // Processing has been delayed until CSn was de-asserted.
+                    rctx.state = RxState::Responding;
+                    ringbuf_entry!(Trace::Responding);
+
+                    if tctx.state == TxState::Idle {
```

Comment:
Is there any way for this condition not to be true?

More generally: it seems like splitting `TxState` and `RxState` up is a mistake.  Having separate enums gives us `N * M` possible state combinations, many of which are nonsensical.  In practice, we want a single unified state machine that's something like
```
enum State {
    Idle,
    RxHeader { count: u32 },
    RxPayload  { count: u32, end: u32 },
    Dispatch,
    TxHeader  { count: u32 },
    TxPayload  { count: u32, end: u32 },
    Invalid,
    Error,
}
```
which gives is `N + M` states instead of `N * M` (and all of the states are meaningful)